src/modules/emote_autocomplete/twitch/EmoteAutocomplete.jsx

```diff
@@ -0,0 +1,161 @@
+  return autocompleteProviders.find(({autocompleteType}) => autocompleteType === 'emote');
+}
+
+function createTwitchEmoteSet(allEmotes) {
+  return {
+    emotes: allEmotes.map((emote) => ({
+      id: serializeEmoteId(emote),
+      modifiers: null,
+      setID: CUSTOM_SET_ID,
+      token: emote.code,
+    })),
+    id: CUSTOM_SET_ID,
+  };
+}
+
+function injectEmoteSets() {
+  const autocompleteEmoteProvider = getAutocompleteEmoteProvider();
+  if (autocompleteEmoteProvider == null) {
+    return;
+  }
+
+  const autocompleteEmotes = autocompleteEmoteProvider.props.emotes;
+
+  const allEmotes = emotes.getEmotesByCategories([
+    EmoteCategories.BETTERTTV_CHANNEL,
+    EmoteCategories.BETTERTTV_GLOBAL,
+    EmoteCategories.BETTERTTV_PERSONAL,
+    EmoteCategories.FRANKERFACEZ_CHANNEL,
+    EmoteCategories.FRANKERFACEZ_GLOBAL,
+  ]);
+
+  const emoteSet = createTwitchEmoteSet(allEmotes);
+  const index = autocompleteEmotes.findIndex(({id}) => id === CUSTOM_SET_ID);
+
+  if (index === -1) {
+    autocompleteEmotes.push(emoteSet);
+  } else {
+    autocompleteEmotes[index] = emoteSet;
+  }
+}
+
+let storeDirtyCallbackCleanup = null;
+let storeUpdatedCallbackCleanup = null;
+let patchImageCallbackCleanup = null;
+let twitchComponentDidUpdate = null;
+export default class EmoteAutocomplete {
+  constructor() {
+    watcher.on('channel.updated', () => this.load());
+    this.load();
+  }
+
+  patchEmoteImage(image, isConnected) {
+    if (!isConnected) {
+      return;
+    }
+
+    const url = String(image.srcset);
+    const deseralizedEmote = deserializeEmoteFromURL(url);
+
+    if (deseralizedEmote == null) {
+      return;
+    }
+
+    const emote = emotes.getEligibleEmote(deseralizedEmote.code);
+    if (emote == null) {
+      return;
+    }
+
+    image.srcset = createSrcSet(emote.images);
+    image.src = emote.images['1x'];
+  }
+
+  load() {
+    const emoteAutocompleteProvider = getAutocompleteEmoteProvider();
+    if (emoteAutocompleteProvider == null) {
+      return;
+    }
+
+    if (emoteAutocompleteProvider.__bttvAutocompletePatched !== PATCHED_SENTINEL) {
+      emoteAutocompleteProvider.__bttvAutocompletePatched = PATCHED_SENTINEL;
+      twitchComponentDidUpdate = emoteAutocompleteProvider.componentDidUpdate;
+
+      // eslint-disable-next-line no-inner-declarations
+      function bttvComponentDidUpdate(prevProps) {
+        if (prevProps.emotes !== this.props.emotes) {
+          injectEmoteSets();
+        }
+
+        if (twitchComponentDidUpdate != null) {
+          twitchComponentDidUpdate.call(this, ...prevProps);
+        }
+      }
+
+      emoteAutocompleteProvider.componentDidUpdate = bttvComponentDidUpdate;
+      emoteAutocompleteProvider.forceUpdate();
+    }
+
+    emoteMenuViewStore.updateEmotes();
+    injectEmoteSets();
+
+    if (storeDirtyCallbackCleanup == null) {
+      storeDirtyCallbackCleanup = emoteMenuViewStore.on('dirty', emoteMenuViewStore.updateEmotes);
```

Comment:
if there is no setting toggle for this, why do we need to store the cleanup at all for the store listeners? couldn't we run these operations in the constructor?

Reply:
additionally, why would we always update the emotes when dirty? this defeats the whole purpose of this dirty optimization, no?